Sources/AsyncHTTPClient/ConnectionPool/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,143 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+final class HTTP1ProxyConnectHandler: ChannelDuplexHandler, RemovableChannelHandler {
+    typealias OutboundIn = Never
+    typealias OutboundOut = HTTPClientRequestPart
+    typealias InboundIn = HTTPClientResponsePart
+
+    enum State {
+        case initialized(EventLoopPromise<Void>)
+        case connectSent(EventLoopPromise<Void>)
+        case headReceived(EventLoopPromise<Void>)
+        case failed(Error)
+        case completed
+    }
+
+    private var state: State
```

Comment:
Do we not want to pull this out into a self-contained state machine?

Reply:
I'm not sure if this is necessary here? @glbrntt @Lukasa opinions?

Reply:
I think either is fine, this looks like a fairly simple state machine. However, with the lack of explicit state machine I think it would be useful to document each state `case` and their valid transitions.